Bootstrap nested fields

oandregal opened the pull request:
Alternative to https://github.com/WordPress/gutenberg/pull/71782

## What 

This PR introduces the ability to have nested fields, by adding a new property `children` in the Field API. It enables a few use cases (leverage the `children` property to declare the type of the array elements, where we now only support strings), but the driving use case is to be able to handle dynamic list of data in DataForm:

By leveraging the children as follows:

```js
{
			id: 'productList',
			label: 'Product List',
			type: 'array',
			children: [
				{
					id: 'name',
					label: 'Name',
					type: 'text',
				},
				{
					id: 'cost',
					label: 'Cost',
					type: 'integer',
				},
				{
					id: 'quantity',
					label: 'Quantity',
					type: 'integer',
				},
			],
		},
```

we can create experiences like this:

https://github.com/user-attachments/assets/85a0957b-a476-4e15-97dd-8218a8b9900e

## Why

We want to operate with dynamic list of data in DataForm.

## Testing instructions

Open storybook (`npm install && npm run storybook:dev`) and visit the stories:

- "DataViews > DataForm > Dynamic Data". This is the main story for dynamic data.
- "DataViews > DataForm > Validation". Updated to highlight how validation support works.
- "DataViews > FieldTypes > array". Updated to highlight how fields with children are rendered by default.

oandregal committed:
Bootstrap nested fields

oandregal committed:
Update claude implementation

oandregal committed:
DataViews: array type renders as UL

oandregal committed:
Provide styles for nested data

oandregal committed:
Add validation story for array with children

oandregal committed:
Separate string and object implementation of array field type definition

oandregal committed:
Validation story

oandregal committed:
Implement add/remove row

oandregal committed:
Add story about dynamic data

oandregal committed:
Better descriptions

oandregal committed:
Update examples: panel

oandregal committed:
Update validation story

oandregal committed:
Document children prop

straku commented:
I think that the new API is flexible and easy to grasp. Combining it with custom `Edit` controls for non-editable fields gives plenty of space to cover a wide set of requirements.

I've noticed that the buttons to add and remove rows are part of the new `TableControl` component. This use case feels specific enough that I wonder whether it is necessary to have something built-in for this purpose or leave this completely in user land, where removing and adding a row would be achieved by editing the data passed to DataForm?

oandregal commented:
> Which of those validation levels are supported now?

All of them.

"Per individual field" + "Per row" can be addressed by implementing a custom function in any of the child fields. Note that `isValid.custom` receives the full item (row). "Per list" can be addressed by implementing `isValid.custom` at the parent field level.

> The main reason why was that the Storybook example that opens in a modal behaves in an unexpected way for me:

I'll fix this.

> Another question I have about extensibility, is around the "Add item" button. I noticed that this string is hardcoded -- is there a way to make it so each list has its own button text, like "Add product" or "Add time slots", depending on the context?
> ...
> This use case feels specific enough that I wonder whether it is necessary to have something built-in for this purpose or leave this completely in user land, where removing and adding a row would be achieved by editing the data passed to DataForm?

I suppose these two things are part of the same: we may want to configure the table so that it can have a fix number of rows (no add/remove actions) + the labels of the actions.

oandregal committed:
Enable delete/add actions to be disabled and modify its labels

oandregal commented:
@straku @jimjasson I've addressed your feedback by allowing consumers to control these actions: they can disable them and update the labels by leveraging the Edit config.

By default, the table displays with delete/add actions and the labels are "Add item", "Remove item":

```js
{
  id: 'productList',
  label: 'Product List',
  type: 'array',
  children: [ /* ... */ ],
}
```

Consumers can configure the labels as such:

```js
{
  id: 'productList',
  label: 'Product List',
  type: 'array',
  Edit: {
    control: 'table',
    actions: {
      delete: "Remove product",
      add: "Add product"
    }
  },
  children: [ /* ... */ ],
}
```

And consumers can also disable the actions (these renders a table without delete/add actions):

```js
{
  id: 'productList',
  label: 'Product List',
  type: 'array',
  Edit: {
    control: 'table',
    actions: {
      delete: fals,
      add: false
    }
  },
  children: [ /* ... */ ],
}
```

We could be even more flexible: change the action icon, adding custom actions, etc. but I think that'd better be a follow-up, only when/if we need it.

oandregal committed:
Split style per component

oandregal committed:
Add table styles

jimjasson commented:
> "Per individual field" + "Per row" can be addressed by implementing a custom function in any of the child fields. Note that isValid.custom receives the full item (row). "Per list" can be addressed by implementing isValid.custom at the parent field level.

Makes sense! 

> I've addressed your feedback by allowing consumers to control these actions: they can disable them and update the labels by leveraging the Edit config.

This is very useful, thanks for addressing it! 

Overall, I like the decision to:
- use the standard combobox element for array fields without children and; 
- the new TableControl for arrays with children (nested or not nested). 

I tested the Fields stories and they all work as expected. 

However, the DataForm story is still not fully functional for me:

(1) When I type in a single character in the product name field, it automatically loses focus:

https://github.com/user-attachments/assets/dd84f743-05c3-4193-aa7a-2621877d439b


(2) The modal doesn't work correctly:


https://github.com/user-attachments/assets/cf9bbb71-df42-45cd-8e70-1f9d43b3c717

oandregal commented:
This is stale, and it has been punted until after WordPress 7.0 beta (Feb 19th).
